src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceCallSite.cs

```diff
@@ -10,6 +10,8 @@ namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
     /// </summary>
     internal abstract class ServiceCallSite
     {
+        private volatile object _value;
```

Comment:
Does it have to be volatile? Even if the initial read won't observe the value we would take the lock and retrieve it.

Reply:
You might be right. I was thinking about accessing the value outside of the lock but if that fails we'll eventually be inside of the lock doing the next read.